lib/drb/drb.rb

```diff
@@ -357,27 +358,67 @@ class DRbConnError < DRbError; end
   # For alternative mechanisms, see DRb::TimerIdConv in drb/timeridconv.rb
   # and DRbNameIdConv in sample/name.rb in the full drb distribution.
   class DRbIdConv
+    MUTEX = Mutex.new
+
+    def initialize
+      @id2ref = {}
+    end
 
     # Convert an object reference id to an object.
-    #
-    # This implementation looks up the reference id in the local object
-    # space and returns the object it refers to.
+    # and returns the object it refers to.
     def to_obj(ref)
-      ObjectSpace._id2ref(ref)
+      _get(ref)
     end
 
     # Convert an object into a reference id.
-    #
-    # This implementation returns the object's __id__ in the local
-    # object space.
     def to_id(obj)
-      case obj
-      when Object
-        obj.nil? ? nil : obj.__id__
-      when BasicObject
-        obj.__id__
+      obj.nil? ? nil : _put(obj)
+    end
+
+    private def _safe_lock
+      Thread.handle_interrupt(Exception => :never) do
+        MUTEX.lock
+        begin
+          yield
+        ensure
+          MUTEX.unlock
```

Comment:
The idiomatic way here would be to use `MUTEX.synchronize(&block)`.  However, maybe there a reason the `Thread.handle_interrupt` and explicit lock/unlock was done instead.  If so, could you add a comment why?

Reply:
Mostly to avoid the overhead of calling the block in synchronize, since this will be hit hard by anyone using a remote object. Is that worth a comment?

Reply:
I would switch to `MUTEX.synchronize(&block)` unless you have a benchmark showing the performance differences are significant enough to warrant this more complex code.  If the performance differences are significant, then keeping the code and adding a comment explaining it seems best.  My uneducated guess would be that the overhead from having a remote object is much higher than the overhead of a mutex synchronization, even on localhost.

Reply:
Fair enough. The block form should be slower on CRuby, but it may be noise compared to the weak map and other remote object overhead.